### PR TITLE
[OpAMP.Client] Add identification frame

### PR DESCRIPTION
**Pull request by RassK**
## Changes

* Adds identification frame (this frame identifies the client for a OpAMP server)

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.OpAmp.Client/Settings/AnyValueUnion.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.OpAmp.Client.Internal.Settings;
+
+namespace OpenTelemetry.OpAmp.Client.Settings;
+
+/// <summary>
+/// Represents a union type that can hold a value of one of several predefined types: integer, boolean, or string.
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/Settings/AnyValueUnion.cs]:
```suggestion
/// Represents a union type that can hold a value of one of several predefined types: integer, double, boolean, or string.
```

File: src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    /// Gets the collection of resources associated with the current instance.
+    /// </summary>
+    public Dictionary<string, AnyValueUnion> NonIdentifyingResources { get; } = [];
+
+    /// <summary>
+    /// Adds an itentifying attribute to the resources collection.
+    /// </summary>
+    /// <param name="key">The unique key associated with the resource.</param>
+    /// <param name="value">The value of the resource to be added.</param>
+    public void AddIdentifyingAttribute(string key, string value)
```
Comment (martincostello) [inline src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs:28]: Do we need both `AddIdentifyingAttribute()` and `AddNonIdentifyingAttribute()`, or could we do something like `AddAttribute(string key, X value, bool identifying)`?

File: test/OpenTelemetry.OpAmp.Client.Tests/AnyValueUnionTests.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+using OpenTelemetry.OpAmp.Client.Internal.Settings;
+using OpenTelemetry.OpAmp.Client.Settings;
+using Xunit;
+
+namespace OpenTelemetry.OpAmp.Client.Tests;
+
+public class AnyValueUnionTests
+{
+    [Fact]
```
Comment (martincostello) [inline test/OpenTelemetry.OpAmp.Client.Tests/AnyValueUnionTests.cs:16]: Add test cases for `null` and equality with itself (e.g. ` Assert.Equal(intValue, intValue);`)

To avoid warnings for those you might need to explicit call the relevant methods, rather than delegating to `Equal()`/`NotEqual()`.

I don't think the xunit helper will give full coverage for `Equals(T)`/`Equals(object)`/`==`/`!=` either.

File: test/OpenTelemetry.OpAmp.Client.Tests/FrameBuilderTests.cs [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+using OpAmp.Proto.V1;
 using OpenTelemetry.OpAmp.Client.Internal;
+using OpenTelemetry.OpAmp.Client.Internal.Services.Heartbeat;
 using Xunit;
 
 namespace OpenTelemetry.OpAmp.Client.Tests;
 
 public class FrameBuilderTests
 {
+    public static IEnumerable<object[]> TestData()
```
Comment (martincostello) [inline test/OpenTelemetry.OpAmp.Client.Tests/FrameBuilderTests.cs]: Use `TheoryData<Func<IFrameBuilder, IFrameBuilder>, Func<AgentToServer, object>>` to strongly-type the test method.

Comment (RassK) [reply test/OpenTelemetry.OpAmp.Client.Tests/FrameBuilderTests.cs]: There is a conflict with accessors. `MemberData` needs to be public, but interfaces and classes referenced there are not public. I can workaround it with `ClassData`, but that leaves `xUnit1044` message without any behavioural changes (since those theories weren't visible before either). At least not a warning.

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/3112/commits/68400e32ab55e3dff537e74b7edcbbc3c16072e1

File: test/OpenTelemetry.OpAmp.Client.Tests/AnyValueUnionTests.cs (hunk 3 above)
Comment (RassK) [reply test/OpenTelemetry.OpAmp.Client.Tests/AnyValueUnionTests.cs:16]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/3112/commits/23ceee8200265e89eb5b45d766906c193bf68dcd

File: src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs (hunk 2 above)
Comment (RassK) [reply src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs:28]: I found that leaves method bodies simpler for us and simpler signatures with more readability to users.

Otherwise no strong opinions.

File: test/OpenTelemetry.OpAmp.Client.Tests/AnyValueUnionTests.cs (hunk 3 above)
Comment (RassK) [reply test/OpenTelemetry.OpAmp.Client.Tests/AnyValueUnionTests.cs:16]: well should cover now all lines

File: src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs (hunk 2 above)
Comment (RassK) [reply src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs:28]: @martincostello @Kielek would we go forward here or does any of you see it as a blocker?

Comment (Kielek) [reply src/OpenTelemetry.OpAmp.Client/Settings/IdentificationSettings.cs:28]: As a user of the library, I would prefer to see  `AddIdentifyingAttribute()` and `AddNonIdentifyingAttribute()` instead of some magic boolean in the code.

@martincostello?
